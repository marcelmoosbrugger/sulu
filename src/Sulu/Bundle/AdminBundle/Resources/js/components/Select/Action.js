// @flow
import React from 'react';
import classnames from 'classnames';
import itemStyles from './selectItem.scss';

export default class Action extends React.PureComponent {
    props: {
        disabled: boolean,
        children?: string,
        onClick: () => void,
        afterAction?: () => void,
    };

    static defaultProps = {
        disabled: false,
    };

    handleButonClick = () => {
        this.props.onClick();
        if (this.props.afterAction) {
            this.props.afterAction();
        }
    };

    render() {
        const classNames = classnames({
            [itemStyles.selectItem]: true,
            [itemStyles.disabled]: this.props.disabled,
        });

        return (
            <li className={classNames}>
                <button onClick={this.handleButonClick} disabled={this.props.disabled}>{this.props.children}</button>
            </li>
        );
    }
}
