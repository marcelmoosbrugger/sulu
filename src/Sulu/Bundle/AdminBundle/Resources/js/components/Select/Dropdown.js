// @flow
import React from 'react';
import dropdownStyles from './dropdown.scss';

export default class Dropdown extends React.PureComponent {
    props: {
        isOpen: boolean,
        children?: React.Element<*>,
    };

    static defaultProps = {
        isOpen: false,
    };

    render() {
        return (
            <ul className={dropdownStyles.dropdown}>
                {this.props.children}
            </ul>
        );
    }
}
