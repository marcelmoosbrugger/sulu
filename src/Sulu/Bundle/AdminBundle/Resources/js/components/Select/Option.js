// @flow
import React from 'react';
import classnames from 'classnames';
import {afterElementsRendered} from '../../services/DOM';
import Icon from '../Icon';
import optionStyles from './option.scss';

const SELECTED_ICON = 'check';

export default class Option extends React.PureComponent {
    props: {
        selected: boolean,
        disabled: boolean,
        focus: boolean,
        value: string,
        children?: string,
        onClick?: (value: string) => void,
    };

    static defaultProps = {
        disabled: false,
        selected: false,
        focus: false,
        value: '',
    };

    item: HTMLElement;
    button: HTMLButtonElement;

    /** @public **/
    getOffsetTop() {
        return this.item.offsetTop;
    }

    handleButtonClick = () => {
        if (this.props.onClick) {
            this.props.onClick(this.props.value);
        }
    };

    setItem = (item: HTMLElement) => this.item = item;
    setButton = (button: ?HTMLButtonElement) => {
        if (!button) {
            return;
        }
        this.button = button;
        if (this.props.focus) {
            afterElementsRendered(() => {
                this.button.focus();
            });
        }
    };

    render() {
        const classNames = classnames({
            [optionStyles.option]: true,
            [optionStyles.selected]: this.props.selected,
        });

        return (
            <li ref={this.setItem}>
                <button
                    className={classNames}
                    ref={this.setButton}
                    onClick={this.handleButtonClick}
                    disabled={this.props.disabled}>
                    {this.props.selected ? <Icon className={optionStyles.icon} name={SELECTED_ICON} /> : ''}
                    {this.props.children}
                </button>
            </li>
        );
    }
}
