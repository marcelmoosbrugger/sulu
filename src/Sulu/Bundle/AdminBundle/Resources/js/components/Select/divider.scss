@import '../../containers/Application/colors.scss';

.divider {
    width: 100%;
    height: 1px;
    background: $frenchGray;
    margin: 5px 0;
}
