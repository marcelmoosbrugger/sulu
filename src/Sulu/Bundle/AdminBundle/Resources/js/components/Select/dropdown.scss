@import '../../containers/Application/colors.scss';

$borderColor: $frenchGray;
$backgroundColor: $wildSand;

.dropdown {
    display: inline-block;
    position: absolute;
    border-radius: 2px;
    padding: 5px 0;
    border: 1px solid $borderColor;
    background: $backgroundColor;
    font-size: 12px;
    list-style-type: none;
    box-shadow: 2px 6px 12px 0 rgba($silver, .22);
    z-index: 1;
    margin: 0 0 0 -15px;
    overflow: auto;
}
