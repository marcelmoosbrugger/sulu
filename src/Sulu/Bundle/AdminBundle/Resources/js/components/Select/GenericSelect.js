// @flow
import React from 'react';
import type {Element, ElementRef} from 'react';
import {action, observable} from 'mobx';
import {observer} from 'mobx-react';
import type {OptionSelectedVisualization, SelectChildren, SelectProps} from './types';
import OverlayList from './OverlayList';
import Action from './Action';
import Option from './Option';
import Label from './Label';
import genericSelectStyles from './genericSelect.scss';

type Props = SelectProps & {
    onSelect: (values: string) => void,
    labelText: string,
    closeOnSelect: boolean,
    isOptionSelected: (option: Element<typeof Option>) => boolean,
    selectedVisualization?: OptionSelectedVisualization,
};

@observer
export default class GenericSelect extends React.PureComponent<Props> {
    static defaultProps = {
        closeOnSelect: true,
    };

    label: ?ElementRef<typeof Label>;
    centeredChildIndex: number;
    @observable isOpen: boolean;

    @action openList = () => {
        this.centeredChildIndex = this.getCenteredChildIndex();
        this.isOpen = true;
    };

    @action closeList = () => {
        this.isOpen = false;
    };

    handleOptionClick = (value: string) => {
        this.props.onSelect(value);
        if (this.props.closeOnSelect) {
            this.closeList();
        }
    };

    handleLabelClick = this.openList;
    handleListRequestClose = this.closeList;
    setLabel = (label: ?ElementRef<typeof Label>) => this.label = label;

    render() {
        const {icon, labelText} = this.props;
        const labelDimensions = this.label ? this.label.getDimensions() : {};
        const listChildren = this.renderListChildren();

        return (
            <div className={genericSelectStyles.select}>
                <Label
                    ref={this.setLabel}
                    icon={icon}
                    onClick={this.handleLabelClick}>
                    {labelText}
                </Label>
                <OverlayList
                    anchorTop={labelDimensions.top}
                    anchorLeft={labelDimensions.left}
                    anchorWidth={labelDimensions.width}
                    anchorHeight={labelDimensions.height}
                    isOpen={this.isOpen}
                    centeredChildIndex={this.centeredChildIndex}
                    onRequestClose={this.handleListRequestClose}>
                    {listChildren}
                </OverlayList>
            </div>
        );
    }

    renderListChildren(): SelectChildren {
        return React.Children.map(this.props.children, (child: any) => {
            if (child.type === Option) {
                child = React.cloneElement(child, {
                    onClick: this.handleOptionClick,
                    selected: this.props.isOptionSelected(child),
                    selectedVisualization: this.props.selectedVisualization,
                });
            }
            if (child.type === Action) {
                child = React.cloneElement(child, {
                    afterAction: this.closeList,
                });
            }
            return child;
        });
    }

    getCenteredChildIndex(): number {
        const index = React.Children.toArray(this.props.children).findIndex(
            (child: any) => child.type === Option && this.props.isOptionSelected(child)
        );
        return index === -1 ? 0 : index;
    }
}
