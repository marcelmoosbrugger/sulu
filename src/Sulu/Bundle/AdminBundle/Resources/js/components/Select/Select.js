// @flow
import React from 'react';
import type {ElementRef} from 'react';
import {action, computed, observable} from 'mobx';
import {observer} from 'mobx-react';
import selectStyles from './select.scss';
import OverlayList from './OverlayList';
import Option from './Option';
import Label from './Label';
import type {SelectChildren} from './types';

type Props = {
    value?: string,
    onChange?: (value: string) => void,
    children: SelectChildren,
    icon?: string,
};

@observer
export default class Select extends React.PureComponent<Props> {
    label: ?ElementRef<typeof Label>;
    @observable isOpen: boolean;

    @action openList = () => {
        this.isOpen = true;
    };

    @action closeList = () => {
        this.isOpen = false;
    };

    @computed get labelText(): string {
        let label = '';
        React.Children.forEach(this.props.children, (child: any) => {
            if (child.type !== Option) {
                return;
            }
            if (!label || this.props.value === child.props.value) {
                label = child.props.children;
            }
        });

        return label;
    }

    handleOptionClick = (value: string) => {
        if (this.props.onChange) {
            this.props.onChange(value);
        }
        this.closeList();
    };

    handleLabelClick = this.openList;
    handleListRequestClose = this.closeList;
    setLabel = (label: ?ElementRef<typeof Label>) => this.label = label;

    render() {
        const labelDimensions = this.label ? this.label.getDimensions() : {};
        const listChildren = this.renderListChildren();
        const centeredChildIndex = this.getCenteredChildIndex(listChildren);

        return (
            <div className={selectStyles.select}>
                <Label
                    ref={this.setLabel}
                    icon={this.props.icon}
                    onClick={this.handleLabelClick}>
                    {this.labelText}
                </Label>
                <OverlayList
                    anchorTop={labelDimensions.top}
                    anchorLeft={labelDimensions.left}
                    anchorWidth={labelDimensions.width}
                    anchorHeight={labelDimensions.height}
                    isOpen={this.isOpen}
                    centeredChildIndex={centeredChildIndex}
                    onRequestClose={this.handleListRequestClose}>
                    {listChildren}
                </OverlayList>
            </div>
        );
    }

    renderListChildren() {
        return React.Children.map(this.props.children, (child: any) => {
            if (child.type === Option) {
                child = React.cloneElement(child, {
                    onClick: this.handleOptionClick,
                    selected: child.props.value === this.props.value && !child.props.disabled,
                });
            }
            return child;
        });
    }

    getCenteredChildIndex(children: any): number {
        const index = React.Children.toArray(children).findIndex((child) => child.props.selected);
        return index === -1 ? 0 : index;
    }
}
