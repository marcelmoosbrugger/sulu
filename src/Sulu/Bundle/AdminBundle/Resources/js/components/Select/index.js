// @flow
import Divider from './Divider';
import Option from './Option';
import Select from './Select';

export {Divider, Option, Select};
