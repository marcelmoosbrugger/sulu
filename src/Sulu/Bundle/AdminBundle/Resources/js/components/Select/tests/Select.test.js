/* eslint-disable flowtype/require-valid-file-annotation */
import {mount} from 'enzyme';
import React from 'react';
import pretty from 'pretty';
import Select from '../Select';
import Divider from '../Divider';
import Option from '../Option';

jest.mock('../../../services/DOM/afterElementsRendered');

afterEach(() => document.body.innerHTML = '');

test('The component should render with the list closed', () => {
    const body = document.body;
    const select = mount(
        <Select>
            <Option value="option-1">Option 1</Option>
            <Option value="option-2">Option 2</Option>
            <Divider />
            <Option value="option-3">Option 3</Option>
        </Select>
    );
    expect(select.render()).toMatchSnapshot();
    expect(body.innerHTML).toBe('');
});

test('The component should render with an icon', () => {
    const body = document.body;
    const select = mount(
        <Select icon="plus">
            <Option value="option-1">Option 1</Option>
            <Option value="option-2">Option 2</Option>
            <Divider />
            <Option value="option-3">Option 3</Option>
        </Select>
    );
    expect(select.render()).toMatchSnapshot();
    expect(body.innerHTML).toBe('');
});

test('The component should render with the correct value', () => {
    const body = document.body;
    const select = mount(
        <Select value="option-2">
            <Option value="option-1">Option 1</Option>
            <Option value="option-2">Option 2</Option>
            <Divider />
            <Option value="option-3">Option 3</Option>
        </Select>
    );
    expect(select.render()).toMatchSnapshot();
    expect(body.innerHTML).toBe('');
});

test('The component should open the list when the label is clicked', () => {
    const body = document.body;
    const select = mount(
        <Select>
            <Option value="option-1">Option 1</Option>
            <Option value="option-2">Option 2</Option>
            <Divider />
            <Option value="option-3">Option 3</Option>
        </Select>
    );
    select.instance().handleLabelClick();
    expect(select.render()).toMatchSnapshot();
    expect(pretty(body.innerHTML)).toMatchSnapshot();
});

test('The component should trigger the change callback and close the list when an option is clicked', () => {
    const body = document.body;
    const onChangeSpy = jest.fn();
    const select = mount(
        <Select onChange={onChangeSpy}>
            <Option value="option-1">Option 1</Option>
            <Option value="option-2">Option 2</Option>
            <Divider />
            <Option value="option-3">Option 3</Option>
        </Select>
    );
    select.instance().handleLabelClick();
    body.getElementsByTagName('button')[2].click();
    expect(onChangeSpy).toHaveBeenCalledWith('option-3');
    expect(body.innerHTML).toBe('');
});
