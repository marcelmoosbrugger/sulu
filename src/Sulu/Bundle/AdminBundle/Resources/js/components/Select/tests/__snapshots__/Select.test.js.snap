// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The component should open the list when the label is clicked 1`] = `
<div
  class="select"
>
  <button
    class="label"
  >
    Option 1
    <span
      aria-hidden="true"
      class="toggle fa fa-chevron-down"
    />
  </button>
  <div>
    
  </div>
</div>
`;

exports[`The component should open the list when the label is clicked 2`] = `
"<div>
  <div data-reactroot=\\"\\" class=\\"container\\">
    <ul class=\\"list\\" style=\\"opacity: 0;\\">
      <li>
        <button class=\\"option\\">
          <!-- react-text: 5 -->
          <!-- /react-text -->
          <!-- react-text: 6 -->Option 1
          <!-- /react-text -->
        </button>
      </li>
      <li>
        <button class=\\"option\\">
          <!-- react-text: 9 -->
          <!-- /react-text -->
          <!-- react-text: 10 -->Option 2
          <!-- /react-text -->
        </button>
      </li>
      <li class=\\"divider\\"></li>
      <li>
        <button class=\\"option\\">
          <!-- react-text: 14 -->
          <!-- /react-text -->
          <!-- react-text: 15 -->Option 3
          <!-- /react-text -->
        </button>
      </li>
    </ul>
  </div>
</div>
<div>
  <div data-reactroot=\\"\\" class=\\"backdrop\\"></div>
</div>"
`;

exports[`The component should render with an icon 1`] = `
<div
  class="select"
>
  <button
    class="label hasIcon"
  >
    <span
      aria-hidden="true"
      class="frontIcon fa fa-plus"
    />
    Option 1
    <span
      aria-hidden="true"
      class="toggle fa fa-chevron-down"
    />
  </button>
  <div>
    
  </div>
</div>
`;

exports[`The component should render with the correct value 1`] = `
<div
  class="select"
>
  <button
    class="label"
  >
    Option 2
    <span
      aria-hidden="true"
      class="toggle fa fa-chevron-down"
    />
  </button>
  <div>
    
  </div>
</div>
`;

exports[`The component should render with the list closed 1`] = `
<div
  class="select"
>
  <button
    class="label"
  >
    Option 1
    <span
      aria-hidden="true"
      class="toggle fa fa-chevron-down"
    />
  </button>
  <div>
    
  </div>
</div>
`;
